### ISSUE TITLE
CLI API?

### ISSUE THREAD

**Issue by kevinSuttle** [postcss/postcss#154]
I'm not finding much in the way of a CLI API, but I'd like to. Goal: to use PostCSS straight from NPM. 
http://blog.keithcirkel.co.uk/how-to-use-npm-as-a-build-tool/

**Comment by kevinSuttle** [comment]
Any progress on this? Would love to start using it straight up.

**Comment by ai** [comment]
Nope. Current plan is:

1. <s>Autoprefixer 5.1</s>
2. <s>postcss-simple-vars</s>
3. <s>postcss-mixins</s>
4. Rewrite docs
5. Plugin guidelines and boilerplate.
6. postcss-cli.
7. Autoprefixer 5.2.
8. PostCSS 5 (async, `postcss.plugin()`, custom parsers)

**Comment by ai** [comment]
I think to remove `autoprefixer` CLI and use `postcss-cli` when it wwill be avaiable. Should we have some `-b` argument in `postcss` to set browsers for any plugins, that uses Browserslist?

**Comment by ai** [comment]
Right now I think about this CLI API:

```
postcss -c CONFIG -p PLUGIN1 -p PLUGIN2 in.css out.css
```

If user will miss `out.css` CLI will print result  to `stdout`. If user will not set any files, CLI will read `stdin` and print to `stdout`.

**Comment by ai** [comment]
Should we use `-p PLUING1 -p PLUGIN2` or `-p PLUGIN1,PLUGIN2`?

**Comment by hawkrives** [comment]
I vote for the separate arg style. 

That also leaves future flexibility for things like subarg syntax – see, for example, how you pass options to Browserify transforms.

**Comment by ai** [comment]
@hawkrives can you show some example of subarg syntax?

**Comment by hawkrives** [comment]
`browserify -t [ babelify --experimental ] -e input.js > output.js`


Where Browserify gets all the arguments outside of the square brackets, and passes the ones inside to the invoked command. 




So the arguments are parsed into something like:




- Browserify

   - `-e` input.js

   - `-t` 

      - babelify

         - `--experimental`




I think substack's pulled the subarg parsing into its own package. (https://github.com/substack/subarg)

**Comment by hawkrives** [comment]
really, github? I'll edit that as soon as I'm back at a computer.

**Comment by lydell** [comment]
+1 for following browserify’s lead.

+1 for removing autoprefixer CLI and replace it with postcss-cli.

autoprefixer CLI can work with many files at once; should postcss-cli too?

**Comment by hawkrives** [comment]
An example for PostCSS would probably look something like `postcss -p [ autoprefixer -b "browsers" ] input.css > output.css`

Note that if you don't want to pass any args to the plugin, you dont need the brackets: `postcss -p autoprefixer input.css > output.css`

**Comment by ai** [comment]
I think to make CLI simpler and allow to pass only one file. We has a `*` in terminal for multiple files (you can set dir as output file and PostCSS will use same file basename).

**Comment by lydell** [comment]
Those two sentences contradict each other. I vote for the latter ;)

**Comment by hawkrives** [comment]
I've got a proposal:

`postcss [-p plugin] [some input files] [-o output-file] [--out-dir dir]`

- Any number of plugins, each with its own `-p` .
- Any number of input files
	- If you give one input file, and no output, it'll write to stdout
	- If you give `-o` with one input file, it writes to that file
	- If you give multiple inputs, and no `--out-dir`, it'll overwrite the input files
	- If you give `--out-dir` with multiple input files, it'll write the corresponding files to the specified directory

I took `--out-dir` from [Babel's cli](https://babeljs.io).

**Comment by ai** [comment]
@hawkrives why we need special API to process many files? We can use bash templates:

```
postcss ./src/*.css ./out/
```

**Comment by hawkrives** [comment]
@ai Because bash expands wildcards before invoking the executable. 

Let's say we have a.css, b.css, and c.css under src. If you run `postcss src/*css output/`, then the arguments that will be passed to the postcss cli will be `src/a.css src/b.css src/c.css outout`. (I'm ignoring the arguments for executable name.)

Right? Or am I misunderstanding something?

**Comment by ai** [comment]
@hawkrives hm, yes you are right. My mistake.

**Comment by pirxpilot** [comment]
I published [postcss-cli](https://github.com/code42day/postcss-cli) module - I need it since we are using Make to build everything.

I think the API is quite close to what's been discussed here. But it does not do stdout/stdin and multiple files. I am sure that can be easily added if people want it.

You are welcome to comment, contribute, clone into postcss org or ignore ;-)

**Comment by ai** [comment]
@pirxpilot nice. But JSON config is a bad idea, because many plugins has a function options.

**Comment by ai** [comment]
I well make big list of issues if you decide to be CLI master after publish a npm module :-D.

**Comment by pirxpilot** [comment]
You're right. I am using yargs so .json comes for free - but if someone needs .js we can easily add it.

**Comment by ai** [comment]
I suggest to remove JSON config from API because many formats can confuse users.

**Comment by MoOx** [comment]
`require(configFile)` offer natively js & json support ;)

**Comment by niksy** [comment]
What @MoOx said, maybe follow the path browserify-shim uses for custom config: https://github.com/thlorenz/browserify-shim#c-config-inside-configshimjs-without-aliases, e.g. load some config by default (`postcssConfig.js`) but leave users with option to override via CLI? This way, other plugins like grunt-postcss can use this config for their `processors` property.

**Comment by ai** [comment]
@pirxpilot I created a issue in your repo. But most of them are not critical. They are just a TODO for future.

Important for me is only:

https://github.com/code42day/postcss-cli/issues/4
https://github.com/code42day/postcss-cli/issues/5
https://github.com/code42day/postcss-cli/issues/2

I added your CLI to PostCSS’s README: 7a5187b

Thanks for nice work :).

**Comment by corysimmons** [comment]
I think this should be reopened and PostCSS should have an official CLI...

- Branding
- Shouldn't have to install 2 tools (you don't `gem install sass; gem install sass-cli`)
- Easy-to-use community agreed upon API
- Community more likely to contribute than on some random project
- Much easier for a newb to get involved in, and explore the ecosystem of, PostCSS if they're able to immediately start using a CLI and some cool plugin they heard about

It's extremely important PostCSS get an official CLI.

It's the difference between installation/usage being:

- `npm i postcss-cli`
- `postcss -p cssnano -w src/css/**/*.css -o dist/css`

And installation/usage being:

- `npm i postcss`
- `postcss -p cssnano -w src/css/**/*.css -o dist/css`
- Oh wait that didn't work... Guess I'll have to read up on this.
- Wtf is [this](https://github.com/postcss/postcss#postcss---)? Where's the CLI?
- Ah okay [here it is](https://github.com/postcss/postcss#runners)
- `npm i postcss-cli`
- `postcss -p cssnano -w src/css/**/*.css -o dist/css`
- Ugh! Why doesn't this work?! Why couldn't they just do it like Sass or Stylus or anything else? Why isn't there an official CLI? Fine whatever I'll read through all the docs on this thing.
- `postcss -u cssnano -d dist/css -w src/**/*.css`
- `throw new TypeError('Non-string provided as watch path: ' + paths);`
- Argh!
- etc.

The point being, this should be a familiar/easy-to-use thing.

I think someone else in this thread mentioned it, but [npm scripts as build tool](http://blog.keithcirkel.co.uk/how-to-use-npm-as-a-build-tool/) is becoming increasingly popular for good reason. They're insanely easy to setup and run in a [plethora of ways](https://github.com/mysticatea/npm-run-all)...

Anyway, just thought I'd whine and cry because PostCSS not having a CLI is a travesty.

I honestly think if you moved away from the whole esoteric marketing scheme of "It does :sparkles: FuTuRe SyNtAx :sparkles: and Autoprefixer :100: was built in it!!!1" and made it insanely easy for newbs to use it'd do wonders for PostCSS' adoption.

**Comment by TrySound** [comment]
@corysimmons Please see https://github.com/sindresorhus/ama/issues/17#issuecomment-117860283 why we opposed to it.

**Comment by ben-eb** [comment]
Separate CLI is a good thing, especially for those who use tools like gulp etc.

https://github.com/svg/svgo/issues/436

**Comment by ai** [comment]
@corysimmons I agree, that we should more think about user experience. I even think that current node.js culture ignores end-user experience and it is bad.

But here is few problems:

* CLI is not a main case of PostCSS usage. Users should use more robust gulp or webpage.
* It is really hard to think and maintain more components.

**Comment by ai** [comment]
@corysimmons what do you think is a main problem of current postcss-cli? Maybe API?

Let's rationalize problem and find better solution.

**Comment by stephenway** [comment]
@ai Let me share my very first experience when my team and I decided to make the switch. We are a clojure/cljs shop which relies on https://clojars.org/ https://www.jfrog.com/artifactory/ and makefile. We did not want to add node to the mix in our apps, so I was tasked with using the unofficial postcss cli in a makefile.
As I began to use the cli, I was finding myself running into a variety of errors that didn't seem to line up with how the gulp/webpack versions tend to work. After about a week or two I was forced to switch to gulp in a separate project.
I believe that the cli needs to work, and should be brought into the postcss org. Being a 3rd party was apart of the barrier to entry as well. I don't think the issue is if it comes with the main postcss project, it just needs some improvement.

**Comment by TrySound** [comment]
@stephenway You can always write issues or send prs. There is a product. To make it good for user we need to know what user want.

**Comment by ai** [comment]
@stephenway I do not like current `postcss-cli` too. Maybe you can create a new better CLI tool. First in seperated test pproject like `postcss-newcli`. And if your CLI will be better, we can change official CLI to your.

I think you should start to collect current CLI problem.

Or maybe you can write this issue to origin CLI author and think together.

**Comment by ben-eb** [comment]
We already have three modules for `@import`, let's not do the same with postcss-cli.

**Comment by stephenway** [comment]
Agreed. I'll work with the maintainer to make it better. I thought providing some perspective would help everyone understand the confusion. For the record, I'm in support of it being separate.

**Comment by pirxpilot** [comment]
**TL;DR** postcss-cli is in postcss org now, needs a maintainer

I wrote postcss-cli to be able to use postcss and autoprefixer with Make. I never intended for it to become an _official_ (whatever _official_ means) CLI runner for postcss. I went in some trouble to [add features to yargs][1] to make handling config files easier. And I [took quite a few patches that added features][2] that - strictly speaking - were only needed if you want to use CLI tool independently from Make (globs, stdin/stdout, Windows compatibility, multiple files, watch functionality etc.)

[I always thought][4] that if someone wanted to take the module in another direction they can fork it and work on it. I never happened. I recently realized that as long as postcss-cli lives in a different organization there might be a reluctance to do that.

So this is what I am going to do:

• I am forking code42day/postcss-cli into postcss organization.
• I am going to update all references to postcss-cli to point to postcss/postcss-cli (if you find any I missed - go for it)
• I am making @aj an owner of the npm module - he'll probably decide to pass the ownership to someone who will maintain it in the long term 

I won't be maintaining postcss/postcss-cli any more. Since I still need a tool  like that, I will be maintaining [postcss-cli-simple][5] that would have a minimum set of features needed to use postcss with make.

Thank you @aj for postcss. And thanks to @FND @schoenwaldnils
@rnons @mkurz @vtrrsl @sciyoshi @drusellers @asimpson who contributed patches to postcss-cli - maybe one of you can take over with postcss-cli maintanence.

[1]: https://github.com/bcoe/yargs/pull/130
[2]: https://github.com/postcss/postcss-cli/blob/master/History.md
[3]: https://www.npmjs.com/package/postcss-cli-simple
[4]: https://github.com/postcss/postcss/issues/154#issuecomment-78419107
[5]: https://www.npmjs.com/package/postcss-cli-simple

**Comment by corysimmons** [comment]
@pirxpilot Thanks for your work on it and thanks for migrating it over to postcss org. :+1: :100:

Everyone else, a separate install is fine. `npm i postcss-cli; postcss -w ...` still looks good.

@ai Please open issues at new repo so we can migrate this conversation over there and decide on an official syntax for it.

**Comment by ai** [comment]
@pirxpilot thanks for work, man :)

**Comment by ai** [comment]
@corysimmons done https://github.com/postcss/postcss-cli/issues/1
